### ISSUE TITLE
reflective tests swallow exceptions thrown by linter rule

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#4262]
**Describe the issue**

If I change the report statement in `avoid_shadowing_type_parameters.dart` from:

```dart
        rule.reportLint(parameter,
            arguments: [parameter.name.lexeme, ancestorKind]);
```

to

```dart
        rule.reportLint(
          parameter, /*arguments: [parameter.name.lexeme, ancestorKind]*/
        );
```

and run `dart run test test/rules/avoid_shadowing_type_parameters_test.dart`, then the test fails, claiming that no lint was reported. If I wrap that in a try/catch and print any thrown exception:

```dart
        try {
          rule.reportLint(
            parameter, /*arguments: [parameter.name.lexeme, ancestorKind]*/
          );
        } catch (e, st) {
          print('caught $e and $st');
        }
```

then sure enough, I see a stack trace:

```dart
caught 'package:analyzer/error/error.dart': Failed assertion: line 104 pos 9: '(arguments ?? const []).length == errorCode.numParameters': Message LintCode.avoid_sh
adowing_type_parameters requires 2 arguments, but 0 arguments were provided and #0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:51:61)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
#2      new AnalysisError (package:analyzer/error/error.dart:104:9)
#3      ErrorReporter.reportErrorForOffset (package:analyzer/error/listener.dart:118:9)
#4      ErrorReporter.reportErrorForNode (package:analyzer/error/listener.dart:102:5)
#5      LintRule.reportLint (package:analyzer/src/lint/linter.dart:686:16)
#6      _Visitor._checkForShadowing (package:linter/src/rules/avoid_shadowing_type_parameters.dart:127:16)
#7      _Visitor._checkAncestorParameters (package:linter/src/rules/avoid_shadowing_type_parameters.dart:98:9)
#8      _Visitor.visitMethodDeclaration (package:linter/src/rules/avoid_shadowing_type_parameters.dart:85:7)
#9      MethodDeclarationImpl.accept (package:analyzer/src/dart/ast/ast.dart:8911:50)
#10     LinterVisitor._runSubscriptions (package:analyzer/src/lint/linter_visitor.dart:1034:14)
#11     LinterVisitor.visitMethodDeclaration (package:analyzer/src/lint/linter_visitor.dart:604:5)
#12     MethodDeclarationImpl.accept (package:analyzer/src/dart/ast/ast.dart:8911:50)
#13     NodeListImpl.accept (package:analyzer/src/dart/ast/ast.dart:9627:20)
#14     EnumDeclarationImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:4303:14)
#15     LinterVisitor.visitEnumDeclaration (package:analyzer/src/lint/linter_visitor.dart:293:10)
#16     EnumDeclarationImpl.accept (package:analyzer/src/dart/ast/ast.dart:4294:50)
#17     NodeListImpl.accept (package:analyzer/src/dart/ast/ast.dart:9627:20)
#18     CompilationUnitImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:2585:21)
#19     LinterVisitor.visitCompilationUnit (package:analyzer/src/lint/linter_visitor.dart:173:10)
#20     CompilationUnitImpl.accept (package:analyzer/src/dart/ast/ast.dart:2578:50)
#21     LibraryAnalyzer._computeLints (package:analyzer/src/dart/analysis/library_analyzer.dart:356:10)
#22     LibraryAnalyzer._computeDiagnostics (package:analyzer/src/dart/analysis/library_analyzer.dart:299:9)
#23     LibraryAnalyzer.analyze (package:analyzer/src/dart/analysis/library_analyzer.dart:92:5)
#24     AnalysisDriver._computeAnalysisResult.<anonymous closure> (package:analyzer/src/dart/analysis/driver.dart:1347:11)
<asynchronous suspension>
#25     PerformanceLog.runAsync (package:analyzer/src/dart/analysis/performance_logger.dart:50:14)
<asynchronous suspension>
#26     AnalysisDriver.performWork (package:analyzer/src/dart/analysis/driver.dart:1055:24)
<asynchronous suspension>
#27     AnalysisDriverScheduler._run (package:analyzer/src/dart/analysis/driver.dart:2146:7)
<asynchronous suspension>
```

**Expected behavior**

I'd hope to be able to see that stack trace regardless. It can result in a lot of lost debugging when the test runner simply claims that a lint wasn't reported.

**Comment by bwilkerson** [comment]
Note that the thrown exception is an `_AssertionError`. I would argue that tests should be run in a mode that enables assertions.

Note also that this is not the only place in the analyzer code bases where assertions are used to report errors that are not expected to occur in production. As a result, this isn't the only kind of test that can silently pass (or fail for less obvious reasons) when asserts are disabled.

**Comment by srawlins** [comment]
Ah good catch, I guess I never think to run `dart run test` with `--enable-asserts` in this package...

**Comment by bwilkerson** [comment]
I have the following in `launch.json` in VS Code, which I think enables asserts when running tests (or anything else) from the IDE:
```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
		"name": "dart",
		"request": "launch",
		"type": "dart",
		"toolArgs": [
			"--enable-asserts"
		],
		"templateFor": "",
	}
    ]
}
```

**Comment by srawlins** [comment]
I think this is settled then. `dart run test` can't show me an assertion error unless I've enabled assertions 🤣  Thanks for the VS Code config!
